### DISCUSSION TITLE
Sofar 12KTL-X-G3 w/ sofar_tlx-g3 profile - CRC Error

### DISCUSSION THREAD

**Discussion by sanderbaks** [davidrapan/ha-solarman#172] [Q&A] [answered]
**Describe the bug**
I own a Sofar 12KTLX-3g
Serial Number of the device: SS2ES112N3U010
I use the Solarman WiFi Logger LSW-3
Firmware: LSW3_15_270A_1.68
Serial Number starts with: 238722

I can Acces my logger using the Deye_String and the Sofar LSW3 files.
All other files show an error.

For both files I see data, but none of them make sense.
I focus on the LSW3 config file.

My Country code keeps changing between Germany and Greece. This also affects the rest of the output.
So, I try to change some register addresses. I focus on 1 to check the results: PV Power 1. In the LSW3 confic file set to 0x000A
Changing it to [0x0586] (taken from the AZZURRO KTL3 config) immediately breaks the config file. Changing it back fixes the connection.

What am I doing Wrong?

**Attach the debug log**

I am not sure what log file to copy or to attach. Please help :)

**To Reproduce**

Just tet it up and change the config files.

**Expected behavior**
I expect the values to reproduce valid numbers.

**Screenshots**
See attached screenshot
<img width="337" alt="Screenshot 2024-09-10 at 09 54 52" src="https://github.com/user-attachments/assets/cf3bbd6d-e8d5-4790-9d74-2a7ba1f3a051">

**Metadata:**
24.09.06

**Comment by davidrapan** [comment]
I would suspect that maybe the register is showing something different than country?

**Comment by sanderbaks** [comment]
Indeed, but if I change a register to another the connections breaks :(

So the problem I face is how do I changes registers in the YAML, without breaking the connection..
It seams so strange.

**Comment by davidrapan** [comment]
Did you tried sofar_tlx-g3.yaml profile? I believe that one is specific for your device. 😉 

Edit: If it won't work. Enable debug logging and attach the file here.

**Comment by sanderbaks** [comment]
Yes I tried all of them, The Sofar TLX does not connect, please see screenshot
<img width="731" alt="Screenshot 2024-09-10 at 11 54 07" src="https://github.com/user-attachments/assets/d23f0233-e31f-434d-83ee-a15cd361cb99">

Essentially none of them connect, except for the Deye String and the Sofar Wifi

**Comment by davidrapan** [comment]
> If it won't work. Enable debug logging and attach the file here.

Please.

**Comment by sanderbaks** [comment]
Here is the Log file..

I really appreciate the work you are doing a lot.

[home-assistant_solarman_2024-09-10T10-29-50.199Z.log](https://github.com/user-attachments/files/16943499/home-assistant_solarman_2024-09-10T10-29-50.199Z.log)

**Comment by davidrapan** [comment]
Hmmmm

@githubDante? Could you please take a look?

**Is it again just about wrong registers?** 🤔

**Comment by githubDante** [comment]
Hi @davidrapan

Probably yes. The response looks like a variation of the resposne in the KeyError issue, but this time the CRC is invalid. The exception code here is 2 (Illegal Data Address).

**Comment by sanderbaks** [comment]
Sounds interesting :). Why does is happen with all the different  config files except 2 ?

**Comment by davidrapan** [comment]
@githubDante, thanks!

> Sounds interesting :). Why does is happen with all the different config files except 2 ?

Cause the file contains different registers, but still not correct ones I'm sure about that as it produces nonsense.

You will need new file which will be based on the sofar_tlx-g3.yaml but with some of the registers left out.

**Comment by sanderbaks** [comment]
But, when I tried to comment out all the registers, it also gave me an error..
is it not the way to proceed ? and find the one register that breaks it?

**Comment by davidrapan** [comment]
The file is very sensitive to proper spacing so you maybe introduced errors to the file itself and that's why it did not work.

So according to the log this 0x0404 - 0x042B range is causing issues so I would focus on registers from that range at first.

**Comment by sanderbaks** [comment]
Awesome, ill start working on this and report back the results!Just curious: is the wifi logger producing the data, and creating the format of it. Or is it the inverter, and is the logger just transferring 1:1? Or even a combination?----------------Sander BaksCommercial & editorial photographerwww.sanderbaks.comwww.hotelcreatives.comOp 11 sep 2024 om 11:20 heeft David Rapan ***@***.***> het volgende geschreven:﻿
The file is very sensitive to proper spacing so you maybe introduced errors to the file itself and that's why it did not work.
So according to the log this 0x0404 - 0x042B range is causing issues so I would focus on registers from that range at first.

**Comment by davidrapan** [comment]
You can also try to add `min_span: 1` into `default:` block so it looks like this:
```
default:
  update_interval: 5
  min_span: 1
  digits: 6
```
That makes it to stop producing register ranges and will ask for specific ones and that will help you locate the issue faster.

Attach log here again after this change and I'll look at it. 😉

Edit: Cause even the registers grouping can cause some issues.

**Comment by davidrapan** [comment]
Logger is formatting the data and uses solarman's proprietary protocol. 😉

**Comment by sanderbaks** [comment]
So, this is my YAML:

# Sofar 3.3-12KTLX-G3
# Note: no modbus energy meter connected

info:
  manufacturer: Sofar
  model: TLX-G3

default:
  update_interval: 5
  min_span: 1
  digits: 6

parameters:
  - group: PV
    items:
      - name: "PV1 Power"
        class: "power"
        state_class: "measurement"
        uom: "W"
        scale: 10
        rule: 1
        registers: [0x0586]
        icon: "mdi:solar-power-variant"


Nothing else.  It breaks...

Using Register 0x000A gives me the value of 30.

Do you guys know any register that should give me a value that I can verify ?

**Comment by davidrapan** [comment]
First of all if you put it here like this in this format I have no way of determining if it's correct at all cause of different formatting.

And if at least the spacing is "matching" with your file then you have it wrong so I'm not surprised it's not working. 😉

**I told you exactly what to do, just add single row with proper spacing according the rest of the file. And it does not matter if it's still not working, I want to see the logs. It's called debugging.** 😉

**Comment by sanderbaks** [comment]
here is the log file :)
[Uploading home-assistant_solarman_2024-09-11T12-56-12.708Z.log…]()

**Comment by davidrapan** [comment]
You did not wait long enough for the file to finish uploading.

**Comment by sanderbaks** [comment]
[home-assistant_solarman_2024-09-11T12-56-12.708Z.log](https://github.com/user-attachments/files/16964328/home-assistant_solarman_2024-09-11T12-56-12.708Z.log)

one more time :)

**Comment by davidrapan** [comment]
So try to remove all entities with registers in this range: 0x0404 - 0x0410

And show me to log again. 😉

**Comment by sanderbaks** [comment]
[home-assistant_solarman_2024-09-11T14-36-00.705Z.log](https://github.com/user-attachments/files/16965209/home-assistant_solarman_2024-09-11T14-36-00.705Z.log)

**Comment by davidrapan** [comment]
You made an error in the profile. Upload the file here and I'll look at it.

**Comment by sanderbaks** [comment]
Here it is! 
[home-assistant_solarman_2024-09-11T15-21-44.795Z.log](https://github.com/user-attachments/files/16965831/home-assistant_solarman_2024-09-11T15-21-44.795Z.log)

**Comment by davidrapan** [comment]
There are still entities w/ registers within the range: 0x040A - 0x0410

**Comment by sanderbaks** [comment]
All gone now :)

[home-assistant_solarman_2024-09-11T15-44-29.566Z.log](https://github.com/user-attachments/files/16966138/home-assistant_solarman_2024-09-11T15-44-29.566Z.log)

**Comment by davidrapan** [comment]
Let's continue: 0x0418

**Comment by sanderbaks** [comment]
You are aware of the fact that the first line "0x0586" already breaks it? That is what I tried to say a this morning.
But here it is!

[home-assistant_solarman_2024-09-11T16-00-46.672Z.log](https://github.com/user-attachments/files/16966333/home-assistant_solarman_2024-09-11T16-00-46.672Z.log)

**Comment by sanderbaks** [comment]
(I mean, don't get me wrong, you know a lot more then I do, but I used to be a programmer myself.  (C, C++,RISC and SQL)

**Comment by sanderbaks** [comment]
But i 10000% appreciate your time and effort to help me!! .. If I can make a donation somewhere ?

**Comment by davidrapan** [comment]
Then you know the drill. Remove all those which are causing issues. 😉 Just don't break the file.

**Comment by sanderbaks** [comment]
Just for my understanding, Should I not be removing from the top down? I am sure that the top one is breaking it. Now I dont know if everything below that would be ok..

Or is that not the way it works? That I am not sure of.

**Comment by sanderbaks** [comment]
Removed all of them one by one, I have the feeling I can not get of the first line.. Will now try from the start, removing from the top until I find the breaking error.

**Comment by davidrapan** [comment]
It always sorts the registers from lowest to highest and then does the requests. So the order in the file does not matter. But if you know what registers breaks it for sure you can remove them, sure. 😉

**Comment by davidrapan** [comment]
But I'm kinda afraid that not even this file has any correct registers set for your device... :-/

**Comment by sanderbaks** [comment]
Ah! Did not know that, that is valuable.
And, I am working the the TLX-G3 file right?
My inverter is the 12KTLX-G3

Just double checking :) none of the registers seem to work.

**Comment by sanderbaks** [comment]
> But I'm kinda afraid that not even this file has any correct registers set for your device... :-/

I agree :(

**Comment by davidrapan** [comment]
We had many examples before that now and then some registers were causing issues for some particular devices and removing them helped but maybe it's not the case with you.

> And, I am working the the TLX-G3 file right?
My inverter is the 12KTLX-G3

Yeah that means it should be file for your device (12K just means 12000 so we are not including the 'K' in the name)

Edit: But maybe the file is just wrong. But you said you did not get any valuable data with any of the files... so that's not good...

**Comment by sanderbaks** [comment]
Its just a strange issue..  I do have connection.. that is the frustrating bit.
Should I try the LAN logger? Its a different Serialnumber.  
If the result is the same we can eliminate that.

**Comment by sanderbaks** [comment]
Anyway, enough for today, I will fiddle on a bit and report back!
Thanks so far! Have a great evening!

**Comment by davidrapan** [comment]
It won't help stick just provides access to the registers and does not sets them.

**Comment by davidrapan** [comment]
Hi @Floppe, I think you added this file right? I'm now looking at it and you named it as KTLX-G3 (stated as such also in the comment in the file) but in the PR you wrote: "Added new Sofar KTLX-G2 inverter." so for which version it is then?

**Comment by Floppe** [comment]
Oh. It should be KTLX-G3.

**Comment by davidrapan** [comment]
Mhm, that's weird then.. It doesn't work for our friend @sanderbaks here...

**Comment by sanderbaks** [comment]
Hi!  It is me again :)

I created a less complex lookup file better be able to see what is happening. I am trying to understand the values I see and to manually match them.  The strange thing is: They change in a funny way. Please see the screen shots. With your knowlegde, what can be the reason for this? The logger can not keep up? In the mean time I did not change anything, I just keep on looking at the screen.



<img width="347" alt="Screenshot 2024-09-12 at 09 37 32" src="https://github.com/user-attachments/assets/8dafb8bd-66b9-428d-9e08-67abb7bacd90">
<img width="366" alt="Screenshot 2024-09-12 at 09 37 48" src="https://github.com/user-attachments/assets/0fb25957-a016-4f75-b0bd-57bfc67688b5">

**Comment by Floppe** [comment]
PV2 voltage insane high on the first screenshot :)

I have the same logger firmware though. Does you logger have good wifi connection? Mine has signal quality only 45% but still works okay.

**Comment by sanderbaks** [comment]
> PV2 voltage insane high on the first screenshot :)
> 
> I have the same logger firmware though. Does you logger have good wifi connection? Mine has signal quality only 45% but still works okay.

The Connection is perfect. It is sitting right next to the access point. It is 100%

I am now trying values according to the attached document, nothing is breaking and I do see values. But I am trying to match them by eye now. Comparing them with the app values. There MUST be a setting that works :)



[SOFARSOLAR ModBus-RTU Communication Protocol.pdf](https://github.com/user-attachments/files/16976307/SOFARSOLAR.ModBus-RTU.Communication.Protocol.pdf)

**Comment by sanderbaks** [comment]
What doe Rule:1 mean ? (ore 2 or 3 in that matter)

**Comment by Floppe** [comment]
It's what kind of value type, unsigned/signed/date/time/raw/bits/version and so on. Don't remember where it was documented though. Perhaps from the original integration.

**Comment by sanderbaks** [comment]
So, you have the same FW of logger, and your register 0x0586 is showing PV Power? This value breaks the integration directly. ..  I think if we solve this, we solve it all. 
I mean, Current is quite stable, Voltage and current  an the AC side is stable.  Is there a way to Search for these values?
Are they integers ? Or does the integration doe any calculations on them? Or conversions?

TEMP is also a nice value to look for I guess? Is that an Integer ? 

Then if i find a few values that match, I can search the internet for a matching list of other registers. And take it from there.

**Comment by Floppe** [comment]
Yes, it shows. Here some statistics for PV1. We have many smalls clouds today here.

![image](https://github.com/user-attachments/assets/cb6f46c8-7063-4479-9fa6-61fdc148717f)

Here very short info how I built the yaml. Google for "sofar g3 external modbus" and you find better documented specs.

From modbus specs

![image](https://github.com/user-attachments/assets/1e58510d-21e3-46ce-94d9-14ab8de37b46)

```
      - name: "PV1 Power"
        class: "power"
        state_class: "measurement"
        uom: "W"
        scale: 10
        rule: 1
        registers: [0x0586]
        icon: "mdi:solar-power-variant"
```

Rule 1 because U16 - unsigned 16 (1 register and only positive values)
scale by 10 so it becomes correct value in Watt.
Or by 0.01 and use kW if you want that.

**Comment by sanderbaks** [comment]
That register directly breaks the whole integration. .. So weird ? What is the difference with my Inverter and yours? The logger is the same FW, so that is not it!

**Comment by sanderbaks** [comment]
Do you know this tool:  https://github.com/Nedel124/Sofar_G3_LSW3

Could this help ?

**Comment by Floppe** [comment]
Well, I have 2 other devices on the same modbus. But I would like to think that have more negative than positive impact on this integration.

I have seen the tool, never tried it out.

**Comment by sanderbaks** [comment]
What inverter do you have ?

See attached software versions .. 
![Schermafbeelding 2024-09-12 om 12 17 20](https://github.com/user-attachments/assets/f7a08f60-e070-4b44-a764-86e09c8af94c)

**Comment by sanderbaks** [comment]
Everywhere I search in the internet the specs say: 0x0586 = Power PV1.
In my case it directly breaks.

I am going crazy!

**Comment by Floppe** [comment]
![Solarman](https://github.com/user-attachments/assets/beb5f423-a3a4-4bd2-9f2f-9c4f11627d3a)
Took a while to get Solarman app working :)

**Comment by sanderbaks** [comment]
Sorry, and to be sure you also have the SOFAR tlx-g3 inverter ?

**Comment by Floppe** [comment]
![image](https://github.com/user-attachments/assets/8da19c5d-8044-4a72-96b2-79bcbb9e8493)

Well, it's actually 8.8. But shouldn't it be the same only that 12 can handle more power.

**Comment by sanderbaks** [comment]
Yes that is correct..  I am going to do a firmware upgrade

**Comment by davidrapan** [comment]
So strange how your screenshots of basic info looks so different... Given the fact it's the same device (when we don't count the rated power).

**Comment by sanderbaks** [comment]
true .. it is driving me mad .. I am trying their support now. But I am not sure if that will help..

**Comment by davidrapan** [comment]
That depends but supports from some of the brands are really trying to be helpful! So maybe you will be lucky! Or maybe the fw upgrade will be enough.

**Comment by sanderbaks** [comment]
So I contacted support. I was very sceptical. But they did a full update, remotely even!
Everything is working now. Execpt for the alerts. but that is a slightly different issue. I will open a new issue for that.

Please see the new version numbers that I have now. and everything is perfectly working.

![Schermafbeelding 2024-09-13 om 12 05 18](https://github.com/user-attachments/assets/26a9659c-8580-439c-81e2-a82d5016bbac)

**Comment by sanderbaks** [comment]
And now also for you guys: @davidrapan @Floppe : You guys are the best.. You can not believe how much I appreciate this!

THANKS!!!

**Comment by davidrapan** [comment]
Hey @sanderbaks, would you be willing to also create some kind of a write up of your issue in discussions for others who could face the same thing? Because I'm sure when this thread is closed it will be lost in time, haha!

**Comment by sanderbaks** [comment]
Yes sure! If you could point me in the direction what it is exactly what you would like i will do it!----------------Sander BaksCommercial & editorial photographerwww.sanderbaks.comwww.hotelcreatives.comOp 17 sep 2024 om 21:13 heeft David Rapan ***@***.***> het volgende geschreven:﻿
Hey @sanderbaks, would you be willing to also create some kind of a write up of your issue in discussions for others who could face the same thing? Cause I'm sure when this thread will be closed it will be lost in time, haha!

**Comment by zkhcj** [comment]
> Hey @sanderbaks, would you be willing to also create some kind of a write up of your issue in discussions for others who could face the same thing? Because I'm sure when this thread is closed it will be lost in time, haha!


how your words were true only 2 days later is amazing 😱

**Comment by sanderbaks** [comment]
I am travelling for work at the moment:)But yes i can do what ever you want, but can you be more specific? What do youMean by write up?----------------Sander BaksCommercial & editorial photographerwww.sanderbaks.comwww.hotelcreatives.comOp 19 sep 2024 om 18:13 heeft zkhcj ***@***.***> het volgende geschreven:﻿

Hey @sanderbaks, would you be willing to also create some kind of a write up of your issue in discussions for others who could face the same thing? Because I'm sure when this thread is closed it will be lost in time, haha!

how your words were true only 2 days later is amazing 😱

**Comment by davidrapan** [comment]
> how your words were true only 2 days later is amazing 😱

That's the experience talking, haha. 😆

> I am travelling for work at the moment:) But yes i can do what ever you want, but can you be more specific? What do you mean by write up?

NP. I was thinking just some kind of a summary of the issue you had and the solution to that. 😉

**Comment by sanderbaks** [comment, marked as answer]
For everybody who will find this and are having the same issues: Not being able to read any info from the inverter. However, I did have a connection. 
Please check the firmware version of you **Inverter**. So NOT the communication Dongle. That will be fine I guess.

Scroll up a few messages and see my before and after. The version numbers have changed. Everything works fine now.
The update was done remotely without me having to do anything.

Contact me anytime if you need help.

Best,

Sander Baks

**Reply by stephencrowther** [reply]
How did you get SOFAR to update your inverter? I've tried contacting them three times and I've had zero response back.

**Reply by zkhcj** [reply]
SOFAR is not really responsive indeed, they didn't upgrade it for me yet.

**Reply by stephencrowther** [reply]
I read that you can upgrade by downloading as from website and putting on usb and doing locally but the instructions say you will need to enter a code half way through which sofar have to supply so still need to contact them for that.

**Reply by stephencrowther** [reply]
I got a reply back from sofar and they just said there wasn’t a firmware update for my KTLM G3 inverter.

I don’t think this is true though as I can download a file from the global site and unzip it and there is a version record in there.

Im on V030006_01_01 which is dated 2022 and it shows latest being V050002_01_02 from March 2024.

**Reply by stephencrowther** [reply]
In case anyone comes across this thread in future - I’ve finally got my inverter working by using this definitions file from this kind person:

https://gist.github.com/gushcs/223ca2c61af2d345238d286746617650

All working fine now with no firmware update needed.

**Comment by lukaszduraj** [comment]
Hit the same CRC / Illegal Data Address behaviour on a Sofar 4.4KTLX-G3 + LSW3 (`LSW3_15_270A_1.32`), inverter SN `SS2ES104M9N584`.

@davidrapan and @githubDante were right that `0x0404–0x042B` is the smoking gun, but the registers themselves are fine. On this stick the **start address** of the request matters:

- The logger only answers holding-register reads whose start is a multiple of `0x20` and which span at most ~40 registers.
- `sofar_g3.yaml` / `sofar_ktlxg3.yaml` start at `0x0404`, `0x0484`, `0x0584`, `0x0684` → the stick returns a malformed frame (`0x90` / exception 2 / bad CRC). That is the CRC error in this thread.
- That also explains why changing PV power from `0x000A` to `0x0586` “broke the connection”: it is not that `0x0586` is invalid, it is that the request no longer starts on a page boundary.

Reads below `0x0400` succeed but are a trap. `0x0000–0x0300` echo the `0x0400` page, so a Deye-style map silently treats `0x0400 = 3` as “Alarm”. The real G3 state is at `0x0404` (`2` = Grid-Connected). Confirmed by reading `0x0445` as ASCII — it spells the inverter serial.

Workaround: keep the native Sofar G3 map, but realign the request windows to `0x0400`, `0x0480`, `0x0580`, `0x0680`.

**Profile**: https://gist.github.com/lukaszduraj/958fd5664acb0812d088e06b62bb4b6f

Drop it into `custom_components/solarman/inverter_definitions/` (or the custom lookup folder) and select it as the profile. No inverter firmware update needed on my unit.

Caveat: the `0x0420` page is not readable this way, so module temperature / insulation resistance from that range are omitted. L3 is included by spanning 37 registers from `0x0480`.

Happy to turn this into a PR if useful. Would be good to hear if it also works on `LSW3_15_270A_1.68` (the firmware from the OP).
